### PR TITLE
Register List Kernels for DML

### PR DESCRIPTION
**Pull request by zhangxiang1993**

### CODE REVIEW

**Comment by PatriceVignola** [comment]
Can you rename the PR to something that doesn't involve the branch name? Also don't forget to change the commit message when merging. #Resolved

File: tensorflow/core/kernels/tensor_array.h [hunk 1]
```diff
@@ -17,6 +17,7 @@ limitations under the License.
 #define TENSORFLOW_CORE_KERNELS_TENSOR_ARRAY_H_
 
 #include <limits.h>
+
```
Comment (PatriceVignola) [inline tensorflow/core/kernels/tensor_array.h]: nit: extra line #Resolved

File: tensorflow/core/kernels/list_kernels.cc [hunk 2]
```diff
@@ -33,6 +31,7 @@ limitations under the License.
 #include "tensorflow/core/lib/core/coding.h"
 #include "tensorflow/core/lib/core/errors.h"
 #include "tensorflow/core/util/util.h"
+#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor" #include "tensorflow/core/framework/op_kernel.h"
```
Comment (PatriceVignola) [inline tensorflow/core/kernels/list_kernels.cc]: Why has it been modified? #Resolved

File: tensorflow/core/kernels/list_kernels.h [hunk 3]
```diff
@@ -753,23 +730,14 @@ class TensorListGather : public OpKernel {
...
+          tensor_array::TensorSetZero<Device, T>(c, &zeros);
         }
-        inputs_flat.emplace_back(new typename TTypes<T, 2>::ConstMatrix(
-            const_cast<const Tensor&>(zeros).shaped<T, 2>(
-                {1, zeros.NumElements()})));
+        values.emplace_back(PersistentTensor(zeros));
       }
     }
-    auto output_flat = output->shaped<T, 2>({1, output->NumElements()});
+    // auto output_flat = output->shaped<T, 2>({1, output->NumElements()});
```
Comment (PatriceVignola) [inline tensorflow/core/kernels/list_kernels.h]: Dead code? #Resolved

File: tensorflow/core/kernels/list_kernels.h [hunk 4]
```diff
@@ -812,9 +780,9 @@ class TensorListFromTensor : public OpKernel {
       // TODO(apassos) maybe not always align; but weird compiler bugs seem to
       // prevent this.
       Tensor aligned;
-      OP_REQUIRES_OK(c, c->allocate_temp(tmp.dtype(), tmp.shape(), &aligned));
-      aligned.flat<T>().device(c->eigen_device<Device>()) =
-          tmp.unaligned_flat<T>();
+      // OP_REQUIRES_OK(c, c->allocate_temp(tmp.dtype(), tmp.shape(),
+      // &aligned));
```
Comment (PatriceVignola) [inline tensorflow/core/kernels/list_kernels.h]: Dead code? #Resolved

File: tensorflow/core/kernels/list_kernels.cc (hunk 2 above)
Comment (jstoecker) [reply tensorflow/core/kernels/list_kernels.cc]: Mistake? Looks like you joined the two lines together. :) #Resolved

**Review by jstoecker — COMMENTED** [review]
@adtsai needs to look at this. The coupling of tensor_array with list_kernels is a little concerning.

File: tensorflow/core/kernels/list_kernels.cc (hunk 2 above)
Comment (zhangxiang1993) [reply tensorflow/core/kernels/list_kernels.cc]: It was modified by the clang-format. But no change here, so I reverted.

---
In reply to: [524980697](https://github.com/microsoft/tensorflow-directml/pull/95#discussion_r524980697) [](ancestors = 524980697)

File: tensorflow/core/kernels/list_kernels.h [hunk 5]
```diff
@@ -798,7 +764,7 @@ class TensorListFromTensor : public OpKernel {
     TensorShape output_shape(t.shape());
     output_shape.RemoveDim(0);
     OP_REQUIRES(c, element_shape.IsCompatibleWith(output_shape),
-                errors::InvalidArgument(
+                errors::InvalidArgument(s
```
Comment (zhangxiang1993) [inline tensorflow/core/kernels/list_kernels.h]: >s [](start = 40, length = 1)

added by mistake, will be reverted #Resolved

File: tensorflow/core/kernels/list_kernels.h [hunk 6]
```diff
@@ -664,9 +642,8 @@ class TensorListSplit : public OpKernel {
       // prevent this.
       Tensor aligned;
       OP_REQUIRES_OK(c, c->allocate_temp(tmp.dtype(), tmp.shape(), &aligned));
-      aligned.flat<T>().device(c->eigen_device<Device>()) =
-          tmp.unaligned_flat<T>();
-      output_list.tensors().emplace_back(aligned);
+      tensor_array::TensorCopyUnaligned<Device, T>(c, &tmp, &aligned);
+      output_list.tensors().emplace_back(tmp);
```
Comment (adtsai) [inline tensorflow/core/kernels/list_kernels.h]: >tmp [](start = 41, length = 3)

Why was this changed? #Resolved

File: tensorflow/core/kernels/list_kernels.h [hunk 7]
```diff
@@ -664,9 +642,8 @@ class TensorListSplit : public OpKernel {
       // prevent this.
       Tensor aligned;
       OP_REQUIRES_OK(c, c->allocate_temp(tmp.dtype(), tmp.shape(), &aligned));
```
Comment (adtsai) [inline tensorflow/core/kernels/list_kernels.h]: >allocate_temp [](start = 27, length = 13)

It looks like TensorCopyUnaligned already calls allocate_temp. Do callers need to allocate their own tensors, or not? #Resolved

Comment (zhangxiang1993) [reply tensorflow/core/kernels/list_kernels.h]: The callers need. Removed the extra line to call allocate_temp.

---
In reply to: [525578893](https://github.com/microsoft/tensorflow-directml/pull/95#discussion_r525578893) [](ancestors = 525578893)

**Comment by zhangxiang1993** [comment]
Thanks for the remind:>

---
In reply to: [728780643](https://github.com/microsoft/tensorflow-directml/pull/95#issuecomment-728780643) [](ancestors = 728780643)

File: tensorflow/core/kernels/list_kernels.h [hunk 8]
```diff
@@ -732,8 +708,9 @@ class TensorListGather : public OpKernel {
       return;
     }
 
-    ConstMatrixVector inputs_flat;
-    inputs_flat.reserve(indices.NumElements());
+    // ConstMatrixVector inputs_flat;
+    std::vector<PersistentTensor> values;
```
Comment (adtsai) [inline tensorflow/core/kernels/list_kernels.h:712]: Dead code?

**Review by adtsai — APPROVED** [review]
:shipit:
